# Network_SNPs
A framework for network analysis and display of SNPs

The group consisted of 6 members. see Abstract for the list.


#Introduction
Our work focuses on using network analysis methods to aid analysis of variant calls. Centrality measure is used as.

# Input

_todo_

# Output
*Filenames:*
_output:_  3-column file with gene names, degrees and betweenness of genes.

                # From this output, it would be smart to focus on CD3EAP since it
                # has the highest centrality and degree measurement
                # Example:
                gene	degrees	btw_centrality
                CD3EAP	500	0.377719224858
                TMC1	30	0.000297257109932
                LRP8	30	0.000768086055561
                SLC30A6	26	0.000223232830101
                ZNF234	22	0.000146138740468
                MYBBP1A	20	0.000310454619974
                HHIPL2	41	0.000495219919278
                CYP2B6	41	0.000465664109724
                CLDN20	17	5.61032381821e-05
_pngs/_: This folder has a figure for each input gene and its neighbor genes.
</p>
_xml_format_: This folder has the xml format of the subnetworks in the pngs folder.
_xml_format_: This folder has the xml format of the subnetworks in the pngs folder.
#Running the pipeline
    testing


#Appendices

##presentation made on day 2 of hackathon
(with additional changes on day 3)
 https://docs.google.com/presentation/d/1PzQrERWxcQQvih6TLuF4IBvmNCFe7D253Qm5UWPAWfA/edit?usp=sharing

